R/expand.R

```diff
@@ -1,92 +1,97 @@
-#' @seealso [complete()] for a common application of `expand`:
-#'   completing a data frame with missing combinations. [expand_grid()]
-#'   is low-level that doesn't deduplicate or sort values.
 #' @export
 #' @examples
-#' library(dplyr)
-#' # All possible combinations of vs & cyl, even those that aren't
-#' # present in the data
-#' expand(mtcars, vs, cyl)
-#'
-#' # Only combinations of vs and cyl that appear in the data
-#' expand(mtcars, nesting(vs, cyl))
-#'
-#' # Implicit missings ---------------------------------------------------------
-#' df <- tibble(
-#'   year   = c(2010, 2010, 2010, 2010, 2012, 2012, 2012),
-#'   qtr    = c(   1,    2,    3,    4,    1,    2,    3),
-#'   return = rnorm(7)
-#' )
-#' df %>% expand(year, qtr)
-#' df %>% expand(year = 2010:2012, qtr)
-#' df %>% expand(year = full_seq(year, 1), qtr)
-#' df %>% complete(year = full_seq(year, 1), qtr)
-#'
-#' # Nesting -------------------------------------------------------------------
-#' # Each person was given one of two treatments, repeated three times
-#' # But some of the replications haven't happened yet, so we have
-#' # incomplete data:
-#' experiment <- tibble(
-#'   name = rep(c("Alex", "Robert", "Sam"), c(3, 2, 1)),
-#'   trt  = rep(c("a", "b", "a"), c(3, 2, 1)),
-#'   rep = c(1, 2, 3, 1, 2, 1),
-#'   measurement_1 = runif(6),
-#'   measurement_2 = runif(6)
+#' # Example data --------------------------------------------------------------
+#' fruits <- tibble(
+#'   type   = c("apple", "orange", "apple", "orange", "orange", "orange"),
+#'   year   = c(2010, 2010, 2012, 2010, 2010, 2012),
+#'   size  =  factor(
+#'     c("XS", "S",  "M", "S", "S", "M"),
+#'     levels = c("XS", "S", "M", "L")
+#'   ),
+#' weights = rnorm(6, as.numeric(size) + 2)
 #' )
 #'
-#' # We can figure out the complete set of data with expand()
-#' # Each person only gets one treatment, so we nest name and trt together:
-#' all <- experiment %>% expand(nesting(name, trt), rep)
+#'
+#' # Expand for all possible combinations --------------------------------------
+#' # Note that all defined, but not necessarily present, levels of the
+#' # factor variable `size` are retained.
+#' fruits %>% expand(type)
+#' fruits %>% expand(type, size)
+#' fruits %>% expand(type, size, year)
+#'
+#'
+#' # Expand for data specific combinations -------------------------------------
+#' fruits %>% expand(nesting(type))
+#' fruits %>% expand(nesting(type, size))
+#' fruits %>% expand(nesting(type, size, year))
+#'
+#'
+#' # Variations on expand ------------------------------------------------------
+#' # Use in conjunction with `forcats::fct_drop` to drop unused factor levels
+#' fruits %>% expand(type, forcats::fct_drop(size), year)
+#
+#' # Use in conjunction with `tidyr::full_seq` to fill in values of continuous variables
+#' fruits %>% expand(type, forcats::fct_drop(size), full_seq(year, 1))
+#' fruits %>% expand(type, forcats::fct_drop(size), 2010:2012)
```

Comment:
Here's the section I think we'd pull to avoid the addition of forcats.

Reply:
removed this section